app/lib/experiments.coffee

```diff
@@ -0,0 +1,51 @@
+$ = require('jqueryify')
+User = require 'zooniverse/lib/models/user'
+Subject = require 'models/subject'
+AnalyticsLogger = require 'lib/analytics'
+
+###
+Define the active experiment here by using a string which exists in http://experiments.zooniverse.org/active_experiments
+If no experiments should be running right now, set this to null, false or ""
+WARNING: using a non-existent experiment will crash the webapp.
+###
+activeExperiment = null
+
+
+currentExperiment = activeExperiment
+currentCohorts = []
+lastFailedAt = null
+
+###
+This will contact the experiment server to find the cohort for this user & subject in the specified experiment
+###
+getCohortRetriever = (user_id = User.current?.zooniverse_id, subject_id = Subject.current?.zooniverseId) ->
+  if typeof activeExperiment=="undefined" || !activeExperiment
+    null
+  else
+    if typeof currentCohorts[currentExperiment] == 'undefined'
+      now = new Date().getTime()
+      if null!=lastFailedAt
+        timeSinceLastFail = now - lastFailedAt.getTime()
+      if null==lastFailedAt || timeSinceLastFail > 30000
+        try
+          $.ajax({
+            url: 'http://experiments.zooniverse.org/experiment/' + currentExperiment + '?userid=' + user_id,
+            dataType: 'json',
+            error: =>
+              lastFailedAt = new Date()
+          }).promise().done( (data, textStatus, jqXHR) =>
+            currentCohorts[currentExperiment] = data.cohort
+            AnalyticsLogger.logEvent 'split'
+          ).fail( (jqXHR, textStatus, errorThrown) =>
+            AnalyticsLogger.logError "500", "Couldn't retrieve experimental split data", "error"
+          )
```

Comment:
Hi Brian, I spent the evening playing with this trying to modify my code to something like your example, I can't get it in a way that works syntactically and doesn't generate an error.

I did improve my understanding a little bit, I learnt that you can only call resolve on a Deferred, not on a Promise, and that $.ajax and $.get only implement Promise not Deferred, ie are not resolvable - I assume this is why you propose creating a separate $.Deferred that is resolvable.

What I can't get to work with your snippets is to return the deferred or indeed its promise as a function, I keep getting an error at the point of trying to call then() on the result, that the then() method doesn't exist.

Let me ask you this: What are the problems you see in the code I have written that your snippets aim to address? Because of course I could throw away what I've written but I'd really like to better understand what you are really telling me with these snippets (ie. to use the expression, I need to learn how to fish rather than just be given a fish!).

Thanks for helping me get my head round this. It's a complex concept, I thought I had it but your review comments suggest maybe I don't fully have it yet! 

(Having said that, my code does produce the desired result... so I guess your concerns are mainly maintainability/efficiency/preferred patterns etc...)